### DISCUSSION TITLE
how different ways of defining same thing work

### DISCUSSION THREAD

**Discussion by sant527** [react-hook-form/react-hook-form#7022] [Q&A] [answered]
I see sometimes

```
      <input
        name="id"
        placeholder="id"
        ref={register({ required: true, maxLength: 50 })}
      />
```

and also

```
      <input
       {...register("id",{ required: true, maxLength: 50 })}
      />
```

I know this is same as

id = {...register("id",{ required: true, maxLength: 50 })}
```
      <input
         name={id.name}
         onChange={id.onChange}
         onBlur={id.onBlur}
         ref={id.ref}
      />
```

But how ` ref={register({ required: true, maxLength: 50 })}`  works

**Comment by bluebill1049** [comment]
refer to the doc: https://react-hook-form.com/api/useform/register and make sure you check the right version that you are using.

**Comment by ritikbanger** [comment, marked as answer]
Hello @sant527 

` ref={register({ required: true, maxLength: 50 })}` is same as `{...register("id",{ required: true, maxLength: 50 })}`

In the first case what we are doing is we are passing a reference of `name` attribute of a particular field to register for registering inputs for validations.

In the second case, we are using the register as a spread operator which will target all values in particular field basically the value given in the `name` attribute. (It is used more when we have more attributes and to explicitly provide the value of name attribute)

` ref={register({ required: true, maxLength: 50 })}` and `{...register("id",{ required: true, maxLength: 50 })}` can be used when you are using 'react-hook-form' with in-built JSX components like `<input/>` 

while you need to use register={register} when giving it as a prop if you design your custom form field, that time ref and spread operator won't work.

Bonus: The type of register is- `register:UseFormRegister<FieldValues>`

**Reply by bluebill1049** [reply]
thank you @ritikbanger for the detailed answer.
